File: js/crc8.js
// CRC-8 estandar (poly 0x07, init 0x00, sin reflejar). Se usa para verificar
// que la secuencia de memes recibida coincide byte a byte con la enviada.
const POLY = 0x07;

export function crc8(bytes) {
  let crc = 0x00;
  for (const byte of bytes) {
    crc ^= byte;
    for (let i = 0; i < 8; i++) {
      crc = (crc & 0x80) ? ((crc << 1) ^ POLY) & 0xFF : (crc << 1) & 0xFF;
    }
  }
  return crc;
}

File: android/app/src/main/java/com/memetransfer/receiver/protocol/Crc8.kt
package com.memetransfer.receiver.protocol

// CRC-8 estandar (poly 0x07, init 0x00, sin reflejar). Puerto directo de
// js/crc8.js - se usa para verificar que la secuencia de memes recibida
// coincide byte a byte con la enviada.
private const val POLY = 0x07

fun crc8(bytes: List<Int>): Int {
    var crc = 0x00
    for (byte in bytes) {
        crc = crc xor (byte and 0xFF)
        repeat(8) {
            crc = if (crc and 0x80 != 0) {
                (crc shl 1) xor POLY
            } else {
                crc shl 1
            }
            crc = crc and 0xFF
        }
    }
    return crc
}
